### PR TITLE
Add GELF support for Promtail.

### LINKED ISSUE #4743: GELF support in promtail

### PR DESCRIPTION
**Pull request by cyriltovena**
**What this PR does / why we need it**:

This PR adds support for GELF UDP protocol to listen and send GELF message to Loki.

**Which issue(s) this PR fixes**:
Fixes #4743

**Special notes for your reviewer**:

**Checklist**
- [x] Documentation added
- [x] Tests updated
- [ ] Add an entry in the `CHANGELOG.md` about the changes.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,7 @@ Release notes for 2.4.1 can be found on the [release notes page](https://grafana
 * [4683](https://github.com/grafana/loki/pull/4683) **owen-d**: Adds replication_factor doc to common config
 * [4681](https://github.com/grafana/loki/pull/4681) **slim-bean**: Loki: check new Read target when initializing boltdb-shipper store
 * [4736](https://github.com/grafana/loki/pull/4736) **sandeepsukhani**: allow applying retention at different interval than compaction
+* [4744](https://github.com/grafana/loki/pull/4744) **cyriltovena**: Promtail: Adds GELF UDP support.
```
Comment (slim-bean) [inline CHANGELOG.md:13]: These got accidentally added in the  2.4.1 release notes, can you move them up above under `main`?

**Review by slim-bean — APPROVED** [review]
LGTM!

File: docs/sources/clients/promtail/configuration.md [hunk 2]
```diff
@@ -983,6 +986,47 @@ The list of labels below are discovered when consuming kafka:
...
+  [ <labelname>: <labelvalue> ... ]
+
+# Whether Promtail should pass on the timestamp from the incoming gelf message.
+# When false, or if no timestamp is present on the gelf message, Promtail will assign the current timestamp to the log when it was processed.
+# Default is false
+use_incoming_timestamp: <bool>
+
+```
+
+**Available Labels:**
```
Comment (james-callahan) [inline docs/sources/clients/promtail/configuration.md:1021]: I'm suprised you don't expose the extra fields as e.g. `__gelf_message_extra_<name>`

Comment (cyriltovena) [reply docs/sources/clients/promtail/configuration.md:1021]: Created an issue https://github.com/grafana/loki/issues/4846

File: clients/pkg/promtail/targets/gelf/gelftarget.go [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+	}()
+}
+
+func (t *Target) handleMessage(msg *gelf.Message) {
+	lb := labels.NewBuilder(nil)
+
+	// Add all labels from the config.
+	for k, v := range t.config.Labels {
+		lb.Set(string(k), string(v))
+	}
```
Comment (jeschkies) [inline clients/pkg/promtail/targets/gelf/gelftarget.go:119]: Isn't this the same for each message? Why is it done each time?

File: docs/sources/clients/promtail/configuration.md [hunk 4]
```diff
@@ -983,6 +986,47 @@ The list of labels below are discovered when consuming kafka:
 
 To keep discovered labels to your logs use the [relabel_configs](#relabel_configs) section.
 
+### GELF
+
+The `gelf` block configures a GELF UDP listener allowing users to push
+logs to Promtail with the [GELF](https://docs.graylog.org/docs/gelf) protocol.
```
Comment (jeschkies) [inline docs/sources/clients/promtail/configuration.md:992]: Does this mean Promtail has a continuous connection with a server? What's happening when it fails?

Comment (cyriltovena) [reply docs/sources/clients/promtail/configuration.md:992]: Promtail is the server.

Comment (cyriltovena) [reply docs/sources/clients/promtail/configuration.md:992]: And there's no connection in UDP.

Comment (jeschkies) [reply docs/sources/clients/promtail/configuration.md:992]: Ah, it's been a long time that I've looked at UDP :see_no_evil:

**Pull request merged**